### DISCUSSION TITLE
How to prevent some actors from clipping

### DISCUSSION THREAD

**Discussion by RafaelLuz** [pyvista/pyvista#3592] [Q&A] [answered]
Hi guys!

I'm trying to simulate a simple Earth - Sun system. Earth is at the origin and also the camera focal point. My goal is to see how the sun changes its position from Earth perspective. However, whenever I put the camera behind Earth and point it to the Sun, Earth is clipped from the scene. It only reappears if the camera is far away from it. Perhaps it is some silly question, but it is driving me crazy. How can I prevent Earth from being clipping?

This is when the camera is too far (my goal is to ger very close to Earth)
![image](https://user-images.githubusercontent.com/12975732/201554155-5f8786d2-0a68-4a83-b52e-0be00ea24a23.png)

one zoom in later and it is completely gone:
![image](https://user-images.githubusercontent.com/12975732/201554215-6fde6106-2743-423a-b19a-6efbeb9a259b.png)

Even when Sun is not in the line of sight, Earth is clipped:
![image](https://user-images.githubusercontent.com/12975732/201554281-0ba364ff-3cbd-45d1-a940-d79d5c75d302.png)

One important detail: if I take the sun out of the scene, I can get arbitrary close to the Earth.

Any help? Thanks in advance!

**Comment by banesullivan** [comment]
This seems like a camera clipping plane issue. What happens if you increase the clipping range with:

```py
pl.camera.clipping_range = (0, 1e7)  # make larger than data extents
```

ref https://docs.pyvista.org/api/core/_autosummary/pyvista.Camera.clipping_range.html#clipping-range

Otherwise, I think we may need the code to generate this scene to really uncover what is going on

**Reply by RafaelLuz** [reply]
I have already tried changing the clipping range as you suggested but it didn't work. So, i'm sending the code.

Thank you very much for you answer and also for you time!!

```python
import numpy

import pyvista
from pyvista import examples


def main():
    # ---------- ---------- ---------- ---------- ---------- plotter
    plotter = pyvista.Plotter()
    plotter.background_color = 'k'
    plotter.enable_anti_aliasing('ssaa')

    # ---------- ---------- ---------- ---------- ---------- earth
    earth_radius = 6378.1  # in km

    earth = pyvista.Sphere(
        radius=earth_radius,
        theta_resolution=120,
        phi_resolution=120,
        start_theta=270.001,
        end_theta=270
    )

    # Initialize the texture coordinates array
    earth.active_t_coords = numpy.zeros((earth.points.shape[0], 2))

    # texture UV-mapping
    for i in range(earth.points.shape[0]):
        earth.active_t_coords[i] = [
            0.5 + numpy.arctan2(-earth.points[i, 0],
                                earth.points[i, 1]) / (2 * numpy.pi),
            0.5 + numpy.arcsin(earth.points[i, 2]
                               / earth_radius) / numpy.pi,
        ]

    earth_texture = examples.load_globe_texture()
    plotter.add_mesh(earth,
                     texture=earth_texture,
                     smooth_shading=True,
                     ambient=0.1,
                     diffuse=10.0,
                     specular=0.5,
                     specular_power=30.0)

    # ---------- ---------- ---------- ---------- ---------- sun
    sun_radius = 695700.0  # in km
    astronomic_unit = 149597870.70000002  # in km

    sun = pyvista.Sphere(
        radius=sun_radius,
        theta_resolution=5,
        phi_resolution=5
    )

    sun.translate((astronomic_unit, 0.0, 0.0), inplace=True)
    # self.sun.translate((2 * self.sun_radius, 0.0, 0.0))
    plotter.add_mesh(sun,
                     emissive=True,
                     style='points_gaussian',
                     render=False,
                     point_size=250)

    # ---------- ---------- ---------- ---------- ---------- camera
    plotter.camera.focal_point = (0.0, 0.0, 0.0)
    plotter.camera.position = (-6.0 * earth_radius, 0.0, 0.0)

    # ---------- ---------- ---------- ---------- ---------- showing
    plotter.show()


if __name__ == '__main__':
    main()
```

**Reply by banesullivan** [reply]
Hm. Yeah this is tough to solve. I'm looking into a similar issue lately and trying to find a better camera and interaction style combination to improve this behavior. I'll come back to this if I uncover anything.

**Reply by Rmelikyan** [reply]
Hi All,

Has anyone thought more about this? I came across a very similar scenario with my own astronomical visualization objectives. Intending to visualize the Earth-Moon system with realistic Moon phases, I included a light point-source at the Sun's location relative to the Earth (using JPL ephemerides) and added a yellow sphere for the sun for fun. The visualized scenario is that of a New Moon from the Earth's perspective such that the Moon is almost directly between the Earth and the Sun.

This is when the clipping problem really started. To diagnose, I set up the camera to be focused on the Moon and initially facing the Earth direction. In this orientation you can zoom pretty deep into the moon before you begin to clip into it. When you start to rotate the scene however, I found that, even with a relatively far camera distance from the moon, the moon quickly began to clip out of the scene. Starting as early as a quarter moon phase, the moon begins to disappear. Eventually all you can see in the background is the Sun out at 1 au.

Searching around I found a couple similar conversations (pyvista/pyvista-support#40, #662, this discussion and a [VTK forum post](https://discourse.slicer.org/t/how-can-i-disable-automatic-reset-of-camera-clipping-range/10839)) and came to the conclusion that the camera clipping range was being automatically updated each time that the camera was moved. So, following some hints from pyvista/pyvista-support#158, I added an observer that reported the current clipping distances in useful units and how far the camera is from the moon. This way I could clearly see that as you rotate the viz towards the solar direction, the inner and outer clipping planes skyrocket and result in the moon blinking from existence.

So, as an interim hack, I modified my callback function to reset the clipping_range at after every camera move. This is chunky, slow, and a bit painful to watch while the scene is moving, but, after the camera stops, I do get the desired scene.

I wonder if there is a way to disable the automatic clipping plane reset during the camera movement so that this type of work can go smoother? The VTK post above suggested that this would mess with the Z-buffer because the inner and outer plane would be too far apart...

Attached is a movie of my scene without and with the clipping range hack.

https://github.com/pyvista/pyvista/assets/57904723/03a81869-42ed-46a4-8b37-5b1210f5b54b

Cheers!

**Reply by RafaelLuz** [reply]
It's nice having an update on this issue since I'm still facing it. 

Can you show me how did you code this workaround? I tried the following, but it didn't work.

```python
def hard_set_clipping_range(*args):
    plotter.camera.clipping_range = (earth_radius, 1*au)

plotter.iren.add_observer(vtk.vtkCommand.EndInteractionEvent, hard_set_clipping_range)

```

**Reply by Rmelikyan** [reply]
Interesting, my code looks like this.

```python
def watching_clipping_range(*args):
    dist_to_moon = norm(np.subtract(ps['moon'].xyz, pl.camera.position))/rMoon # in lunar radii
    pl.add_text(f"Camera Dist to Moon: {dist_to_moon:.1f} (Rmoon)", name="ldist", color="white", position="upper_left")
    pl.add_text(f"Inner Clipping Range: {pl.camera.clipping_range[0]/rMoon:.1f} (Rmoon)", name="inClip", color="white", position="upper_edge")
    pl.add_text(f"Outer Clipping Range: {pl.camera.clipping_range[1]/AU:.3f} (AU)", name="outClip", color="white", position="upper_right")
    pl.add_text("", name="Newldist", color="white", position="lower_left")
    pl.add_text("", name="NewinClip", color="white", position="lower_edge")
    pl.add_text("", name="NewoutClip", color="white", position="lower_right")

def hard_set_clipping_range(*args):
    dist_to_moon = norm(np.subtract(ps['moon'].xyz, pl.camera.position))/rMoon # in lunar radii
    pl.add_text(f"Camera Dist to Moon: {dist_to_moon:.1f} (Rmoon)", name="ldist", color="white", position="upper_left")
    pl.add_text(f"Inner Clipping Range: {pl.camera.clipping_range[0]/rMoon:.1f} (Rmoon)", name="inClip", color="white", position="upper_edge")
    pl.add_text(f"Outer Clipping Range: {pl.camera.clipping_range[1]/AU:.3f} (AU)", name="outClip", color="white", position="upper_right")
    pl.camera.clipping_range = (0.1*rMoon, AU)
    pl.add_text(f"New Camera Dist to Moon: {dist_to_moon:.1f} (Rmoon)", name="Newldist", color="white", position="lower_left")
    pl.add_text(f"New Inner Clipping Range: {pl.camera.clipping_range[0]/rMoon:.1f} (Rmoon)", name="NewinClip", color="white", position="lower_edge")
    pl.add_text(f"New Outer Clipping Range: {pl.camera.clipping_range[1]/AU:.3f} (AU)", name="NewoutClip", color="white", position="lower_right")
    return

def set_callback_method(flag=False):
    if flag:
        pl.iren.remove_observer(curr_observer["obs"])
        curr_observer["obs"] = pl.iren.add_observer(vtk.vtkCommand.EndInteractionEvent, hard_set_clipping_range)
        return
    if "obs" in list(curr_observer.keys()):
        pl.iren.remove_observer(curr_observer["obs"])
    curr_observer["obs"] = pl.iren.add_observer(vtk.vtkCommand.EndInteractionEvent, watching_clipping_range)

curr_observer = {}
curr_observer["obs"] = set_callback_method()
pl.add_checkbox_button_widget(set_callback_method, value=False, position=[10, 100])
pl.add_text("Hard Set Clipping Range", [65, 100 + 5], color="white")
```

which more or less reduces down to exactly as you have above if you take out the button toggle and just focus on the hard_set_clipping_range without the text details.

With the above code, as you may have noticed in the video, my bodies are still clipping (even with the hard set clipping range) while the camera is moving or zooming (presumably because vtk is resetting the clipping range throughout the entire camera movement) but the bodies return to their correct view when I stop moving the camera.

**Reply by banesullivan** [reply, marked as answer]
I remember disabling the camera clipping reset before. I'm pretty sure it was done on the interactor style via `vtkInteractorStyle.AutoAdjustCameraClippingRangeOff`

ref https://vtk.org/doc/nightly/html/classvtkInteractorStyle.html#a78df51732139037fbf3053ce2c08ad40

Here's a quick example that I verified:

```py
import pyvista as pv
import vtk

pl = pv.Plotter()
pl.add_mesh(pv.Cone())

def callback(*args):
    pl.add_text(f"{pl.camera.clipping_range}", name="clipping range")

pl.iren.add_observer(vtk.vtkCommand.EndInteractionEvent, callback)

pl.iren.style.AutoAdjustCameraClippingRangeOff()

pl.show()
```

IMO, we should more conveniently expose this in pyvista

**Reply by RafaelLuz** [reply]
Hey @banesullivan ! Thank you very much. I just implemented and it worked like a charm! 

![image](https://github.com/pyvista/pyvista/assets/12975732/15147e58-26ad-48a5-beae-58ffcd7d04d5)
